pkg/trace/writer/sender.go

```diff
@@ -210,29 +207,28 @@ outer:
 
 // Push pushes p onto the sender's queue, to be written to the destination.
 func (s *sender) Push(p *payload) {
-	for {
-		select {
-		case s.queue <- p:
-			// ok
-			s.inflight.Inc()
-			return
-		default:
-			// drop the oldest item in the queue to make room
-			select {
-			case p := <-s.queue:
-				s.releasePayload(p, eventTypeDropped, &eventData{
-					bytes: p.body.Len(),
-					count: 1,
-				})
-			default:
-				// the queue got drained; not very likely to happen, but
-				// we shouldn't risk a deadlock
-				continue
-			}
-		}
+	s.mu.RLock()
+	if s.closed {
+		s.mu.RUnlock()
+		return
+	}
+	s.mu.RUnlock()
+	select {
+	case s.queue <- p:
+	default:
+		metrics.Count("datadog.trace_agent.sender.push_blocked", 1, nil, 1)
+		s.queue <- p
 	}
+	s.inflight.Inc()
 }
 
+// maxRetries is the maximum number of retries that the sender will perform
+// before giving up. Note that the sender may not perform all maxRetries if
+// the agent is under load and the outgoing payload queue is full. In that
+// case, the sender will drop failed payloads when it is unable to enqueue
+// them for another retry.
+const maxRetries = 4
```

Comment:
Is there a good reason to set it to 4? Should we also make it configurable?

Reply:
I'm on the fence about this one. 4 is just a magic number chosen basically out of nowhere, so I'm happy to change it. 

OTOH, I would like to avoid adding more configs. This probably has an optimal value or range of values, and I don't think users should need to change it. The fewer knobs for users to turn, the better, not because we don't want them to have options, but because we don't want them to *have* to tune every parameter to get good behavior.

If we're not happy with 4, maybe we can experiment or discuss and find a better value.

Reply:
Having the option to change the value can be helpful to mitigate regressions without having to release a patch. That's the main reason behind the suggestion.

Reply:
:+1: That makes sense. Maybe we can have a config but keep it undocumented for now, perhaps removing it later. WDYT?

Reply:
Added an undocumented config for this. We can remove it later.